### PR TITLE
[SYCL-MLIR]: Remove cgeist hack used to represent homogenous structs as arrays

### PR DESCRIPTION
**Pull request by etiotto**
This PR fixes issue [#7662](https://github.com/intel/llvm/issues/7662) and remove the code in `cgeist` that represent a struct containing homogenous member data as an array. This kind of "optimizations" is better done as a transformation pass (or we might get it by abiding to the target ABI which in some cases flattens a struct).

### CODE REVIEW

File: polygeist/tools/cgeist/Test/Verification/arrayconsmemrefinner.cpp [hunk 1]
```diff
@@ -29,13 +29,13 @@ void kern() {
...
 // CHECK-NEXT:       %2 = arith.index_cast %arg1 : index to i64
 // CHECK-NEXT:       %3 = llvm.getelementptr %1[%2] : (!llvm.ptr<struct<(i32)>>, i64) -> !llvm.ptr<struct<(i32)>>
-// CHECK-NEXT:       %4 = "polygeist.pointer2memref"(%3) : (!llvm.ptr<struct<(i32)>>) -> memref<?x1xi32>
-// CHECK-NEXT:       call @_ZN11AIntDividerC1Ev(%4) : (memref<?x1xi32>) -> ()
+// CHECK-NEXT:       call @_ZN11AIntDividerC1Ev(%3) : (!llvm.ptr<struct<(i32)>>) -> ()
 // CHECK-NEXT:     }
 // CHECK-NEXT:     return
 // CHECK-NEXT:   }
-// CHECK:   func @_ZN11AIntDividerC1Ev(%arg0: memref<?x1xi32>) attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
+// CHECK:   func.func @_ZN11AIntDividerC1Ev(%arg0: !llvm.ptr<struct<(i32)>>)
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/arrayconsmemrefinner.cpp:36]:
```suggestion
// CHECK:   func.func @_ZN11AIntDividerC1Ev(%arg0: !llvm.ptr<struct<(i32)>>)  attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
```

File: polygeist/tools/cgeist/Test/Verification/classrefmem.cpp [hunk 2]
```diff
@@ -23,12 +23,13 @@ void Q(A& a) {
...
-// CHECK:   func @_ZN1A3addEv(%arg0: memref<?x1xmemref<?xi32>>) attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
-// CHECK-NEXT:     %0 = affine.load %arg0[0, 0] : memref<?x1xmemref<?xi32>>
-// CHECK-NEXT:     call @_Z4oaddRi(%0) : (memref<?xi32>) -> ()
+// CHECK:   func @_ZN1A3addEv(%arg0: !llvm.ptr<!llvm.struct<(memref<?xi32>)>>) attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
+// CHECK-NEXT:     %0 = llvm.getelementptr %arg0[0, 0] : (!llvm.ptr<!llvm.struct<(memref<?xi32>)>>) -> !llvm.ptr<memref<?xi32>>
+// CHECK-NEXT:     %1 = llvm.load %0 : !llvm.ptr<memref<?xi32>>
+// CHECK-NEXT:     call @_Z4oaddRi(%1) : (memref<?xi32>) -> ()
 // CHECK-NEXT:     return
-// CHECK-NEXT:   }
+// CHECK-NEXT:   }
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/classrefmem.cpp:35]:
```suggestion
// CHECK-NEXT:   }
```

File: polygeist/tools/cgeist/Test/Verification/gettimeofday.c [hunk 3]
```diff
@@ -8,19 +8,19 @@ double alloc() {
...
 
-// clang-format off
 // CHECK:   func @alloc() -> f64
 // CHECK-NEXT:     %cst = arith.constant 9.9999999999999995E-7 : f64
-// CHECK-NEXT:     %alloca = memref.alloca() : memref<1x2xi64>
-// CHECK-NEXT:     %0 = "polygeist.memref2pointer"(%alloca) : (memref<1x2xi64>) -> !llvm.ptr<struct<(i64, i64)>>
-// CHECK-NEXT:     %1 = llvm.mlir.null : !llvm.ptr<i8>
+// CHECK-NEXT:     %c1_i64 = arith.constant 1 : i64
+// CHECK-NEXT:     %0 = llvm.alloca %c1_i64 x !llvm.struct<(i64, i64)> : (i64) -> !llvm.ptr<struct<(i64, i64)>>
+// CHECK-NEXT:     %1 = llvm.mlir.null : !llvm.ptr<struct<(i32, i32)>>
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/gettimeofday.c:15]: on the buildbot machine, this line expects to be `%1 = llvm.mlir.null : !llvm.ptr<i8>`
```suggestion
// CHECK-NEXT:     %1 = llvm.mlir.null : !llvm.ptr<{{.*}}>
```

File: polygeist/tools/cgeist/Test/Verification/ident.cpp [hunk 4]
```diff
@@ -41,54 +41,58 @@ void lt_kernel_cuda(MTensorIterator& iter) {
...
+// CHECK-NEXT:       %4 = llvm.getelementptr %1[0, 0] : (!llvm.ptr<struct<(ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>)>>) -> !llvm.ptr<ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>>
+// CHECK-NEXT:       llvm.store %arg0, %4 : !llvm.ptr<ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>>
+// CHECK-NEXT:       %5 = llvm.load %1 : !llvm.ptr<struct<(ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>)>>
+// CHECK-NEXT:       llvm.store %5, %0 : !llvm.ptr<struct<(ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>)>>
+// CHECK-NEXT:       call @_ZZ14lt_kernel_cudaENK3$_0clEv(%0) : (!llvm.ptr<struct<(ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>)>>) -> ()
 // CHECK-NEXT:     }
 // CHECK-NEXT:     return
 // CHECK-NEXT:   }
-// CHECK:   func @_ZNK15MTensorIterator11input_dtypeEv(%arg0: !llvm.ptr<!llvm.struct<(!llvm.struct<(memref<?x2xi8>)>)>>) -> i8 attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
+// CHECK:   func.func @_ZNK15MTensorIterator11input_dtypeEv(%arg0: !llvm.ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>) -> i8 attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/ident.cpp:60]:
```suggestion
// CHECK:   func.func @_ZNK15MTensorIterator11input_dtypeEv(%arg0: !llvm.ptr<struct<(struct<(ptr<struct<(i8, i8)>>)>)>>) -> i8 attributes {llvm.linkage = #llvm.linkage<linkonce_odr>} {
```

File: polygeist/tools/cgeist/Test/Verification/new.cpp [hunk 5]
```diff
@@ -9,12 +9,30 @@ struct A {
...
+  // CHECK-NEXT:   llvm.store %cst_0, %4 : !llvm.ptr<f32>
+  // CHECK-NEXT:   %5 = llvm.getelementptr %1[0, 1] : (!llvm.ptr<struct<(f32, f32)>>) -> !llvm.ptr<f32>
+  // CHECK-NEXT:   llvm.store %cst, %5 : !llvm.ptr<f32>
+  // CHECK-NEXT:   %6 = llvm.load %1 : !llvm.ptr<struct<(f32, f32)>>
+  // CHECK-NEXT:   llvm.store %6, %0 : !llvm.ptr<struct<(f32, f32)>>
+  // CHECK-NEXT:   %7 = call @_ZN1AaSEOS_(%3, %0) : (!llvm.ptr<struct<(f32, f32)>>, !llvm.ptr<struct<(f32, f32)>>) -> !llvm.ptr<struct<(f32, f32)>>
+  // CHECK-NEXT:   call @_Z1fP1A(%3) : (!llvm.ptr<struct<(f32, f32)>>) -> ()
+  // CHECK-NEXT:   return %c0_i32 : i32
+
+  // BUG: cgeist does not initialize the new struct (issue #).
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/new.cpp:31]: Please open an issue.

File: polygeist/tools/cgeist/Test/Verification/stream.cu [hunk 6]
```diff
@@ -14,8 +14,8 @@ void run(cudaStream_t stream1, int *array, int n) {
 }
 
 // CHECK:   func.func @_Z3runP10cudaStreamPii(%arg0: !llvm.ptr<struct<()>>, %arg1: memref<?xi32>, %arg2: i32) attributes {llvm.linkage = #llvm.linkage<external>} {
-// CHECK:          %12 = "polygeist.stream2token"(%arg0) : (!llvm.ptr<struct<()>>) -> !gpu.async.token
-// CHECK-NEXT:     %13 = gpu.launch async [%12] blocks(%arg3, %arg4, %arg5) in (%arg9 = %1, %arg10 = %3, %arg11 = %5) threads(%arg6, %arg7, %arg8) in (%arg12 = %7, %arg13 = %9, %arg14 = %11) {
+// CHECK:          %26 = "polygeist.stream2token"(%arg0) : (!llvm.ptr<struct<()>>) -> !gpu.async.token
+// CHECK-NEXT:     [[LAUNCH_RET:.*]] = gpu.launch async [%26] blocks(%arg3, %arg4, %arg5) in (%arg9 = {{.*}}, %arg10 = {{.*}}, %arg11 = {{.*}}) threads(%arg6, %arg7, %arg8) in (%arg12 = {{.*}}, %arg13 = {{.*}}, %arg14 = {{.*}}) {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/stream.cu:18]: `[[LAUNCH_RET]]` is not used.
```suggestion
// CHECK-NEXT:     %27 = gpu.launch async [%26] blocks(%arg3, %arg4, %arg5) in (%arg9 = {{.*}}, %arg10 = {{.*}}, %arg11 = {{.*}}) threads(%arg6, %arg7, %arg8) in (%arg12 = {{.*}}, %arg13 = {{.*}}, %arg14 = {{.*}}) {
```

File: polygeist/tools/cgeist/Test/Verification/arrayconsmemref.cpp [hunk 7]
```diff
@@ -13,16 +13,19 @@ void kern() {
 // CHECK-DAG:     %c25 = arith.constant 25 : index
 // CHECK-DAG:     %c1 = arith.constant 1 : index
 // CHECK-DAG:     %c0 = arith.constant 0 : index
-// CHECK-NEXT:     %alloca = memref.alloca() : memref<25x1xi32>
-// CHECK-NEXT:     scf.for %arg0 = %c0 to %c25 step %c1 {
-// CHECK-NEXT:       %0 = "polygeist.subindex"(%alloca, %arg0) : (memref<25x1xi32>, index) -> memref<?x1xi32>
-// CHECK-NEXT:       call @_ZN11AIntDividerC1Ev(%0) : (memref<?x1xi32>) -> ()
+// CHECK-DAG:     %c1_i64 = arith.constant 1 : i64
+// CHECK-NEXT:    llvm.alloca %c1_i64 x !llvm.array<25 x struct<(i32)>> : (i64) -> !llvm.ptr<array<25 x struct<(i32)>>>
+// CHECK-NEXT:    %1 = llvm.getelementptr %0[0, 0] : (!llvm.ptr<array<25 x struct<(i32)>>>) -> !llvm.ptr<struct<(i32)>>
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/arrayconsmemref.cpp:18]:
```suggestion
// CHECK-DAG:      %c1_i64 = arith.constant 1 : i64
// CHECK-NEXT:     llvm.alloca %c1_i64 x !llvm.array<25 x struct<(i32)>> : (i64) -> !llvm.ptr<array<25 x struct<(i32)>>>
// CHECK-NEXT:     %1 = llvm.getelementptr %0[0, 0] : (!llvm.ptr<array<25 x struct<(i32)>>>) -> !llvm.ptr<struct<(i32)>>
```

File: polygeist/tools/cgeist/Test/Verification/pair.c [hunk 8]
```diff
@@ -18,25 +18,30 @@ int create() {
...
 }
 
-// CHECK:   func @byval(%arg0: memref<?x2xi32>, %arg1: i32, %arg2: memref<?x2xi32>)
-// CHECK-NEXT:     affine.store %arg1, %arg0[0, 1] : memref<?x2xi32>
-// CHECK-NEXT:     %0 = affine.load %arg0[0, 0] : memref<?x2xi32>
-// CHECK-NEXT:     affine.store %0, %arg2[0, 0] : memref<?x2xi32>
-// CHECK-NEXT:     %1 = affine.load %arg0[0, 1] : memref<?x2xi32>
-// CHECK-NEXT:     affine.store %1, %arg2[0, 1] : memref<?x2xi32>
-// CHECK-NEXT:     return
+// CHECK:    func.func @byval(%arg0: !llvm.struct<(i32, i32)>, %arg1: i32) -> !llvm.struct<(i32, i32)>
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/pair.c:21]: Interesting, the original code put the return value as the last argument, and this change changes that behavior.

Comment (victor-eds) [reply polygeist/tools/cgeist/Test/Verification/pair.c:21]: `clang` returns `i64` instead of the pair. Does this violate the calling convention?

Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/pair.c:21]: We do not yet follow the ABI for struct returns. Compiling the test with -w, we can see a couple of warnings: 
```
cgeist -O0 --function='*' -S /localdisk2/etiotto/intel-llvm/polygeist/tools/cgeist/Test/Verification/pair.c
warning: struct should be flattened but MLIR codegen cannot yet handle it. Needs to be fixed.
warning: struct should be flattened but MLIR codegen cannot yet handle it. Needs to be fixed.
```
I think once we implement this part of the ABI correctly, we will get the LLVM IR clang generates (ABI conforming).

File: polygeist/tools/cgeist/Test/Verification/new.cpp (hunk 5 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/new.cpp]: Done. https://github.com/intel/llvm/issues/7709

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/tools/cgeist/Test/Verification/pair.c (hunk 8 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Test/Verification/pair.c:21]: Nice. Thanks, Ettore.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**